apps/rush-lib/src/utilities/install-rush-for-ci.ts

```diff
@@ -0,0 +1,173 @@
+    // We're on Windows
+    const whereOutput: string = childProcess.execSync('where npm', { stdio: [] }).toString();
+    const lines: string[] = whereOutput.split(os.EOL).filter((line) => !!line);
+    npmPath = lines[lines.length - 1];
+  } else {
+    // We aren't on Windows - assume we're on *NIX or Darwin
+    npmPath = childProcess.execSync('which npm', { stdio: [] }).toString();
+  }
+} catch (e) {
+  console.error(`Unable to determine the path to the NPM tool: ${e}`);
+  process.exit(1);
+}
+
+npmPath = npmPath.trim();
+console.log(os.EOL + `NPM executable is '${npmPath}'`);
+
+if (!fs.existsSync(npmPath)) {
+  console.error('The NPM executable does not exist');
+  process.exit(1);
+}
+
+const rushPathParts: string[] = ['common', 'temp', 'local-rush'];
+let rushPath: string = rushJsonDirectory;
+for (const rushPathPart of rushPathParts) {
+  rushPath = path.join(rushPath, rushPathPart);
+  try {
+    if (!fs.existsSync(rushPath)) {
+      fs.mkdirSync(rushPath);
+    }
+  } catch (e) {
+    console.error(`Error building local rush installation directory: ${e}`);
+    process.exit(1);
+  }
+}
+
+console.log(os.EOL + `Expected Rush version is ${expectedVersion}`);
+
+// Check for the Rush version
+let installedVersion: string = undefined!;
+let installedVersionValid: boolean = false;
+try {
+  const spawnResult: childProcess.SpawnSyncReturns<Buffer> = childProcess.spawnSync(
+    npmPath, ['list', PACKAGE_NAME, 'version'],
+    { cwd: rushPath, stdio: ['pipe', 'pipe', 'pipe'] }
+  );
+  const output: string = spawnResult.output.toString();
+  const matches: string[] | null = /@microsoft\/rush\@([0-9a-zA-Z.+\-]+)/.exec(output);
+  // If NPM finds the wrong version in node_modules, that version will be in matches[1].
+  // But if it's not installed at all, then NPM instead uselessly tells us all about
+  // the version that we DON'T have ('missing:')
+  if (matches && matches.length === 2 && !output.match(/missing\:/g)) {
+    installedVersion = matches[1]!;
+
+    if (spawnResult.status === 0) {
+      installedVersionValid = true;
+    }
+  }
+} catch (error) {
+  // (this happens if we didn't find the installed package)
+}
+
+if (installedVersion) {
+  console.log(os.EOL + `Installed version is ${installedVersion}`);
+} else {
+  console.log(os.EOL + 'Rush does not appear to be installed');
+}
+
+if (!installedVersionValid || installedVersion !== expectedVersion) {
+  const npmrcPath: string = path.join(rushJsonDirectory, 'common', 'config', 'rush', '.npmrc');
+  const rushNpmrcPath: string = path.join(rushPath, '.npmrc');
+  if (fs.existsSync(npmrcPath)) {
+    try {
+      const npmrcFileLines: string[] = fs.readFileSync(npmrcPath).toString().split('\n').map((line) => line.trim());
+      const resultLines: string[] = [];
+      // Trim out lines that reference environment variables that aren't defined
+      for (const line of npmrcFileLines) {
+        const environmentVariables: string[] | null = line.match(/\$\{([^\}]+)\}/g);
```

Comment:
>\} [](start = 73, length = 2)

Should we insist that it's a simple environment variable?  For example this might match `${FOO_${BAR}}`.

Reply:
We use [this RegExp](https://github.com/Microsoft/web-build-tools/blob/a29b979dac974add0d7d4a09cb70ae197f0b6de4/libraries/ts-command-line/src/CommandLineParameter.ts#L46) in ts-command-line:


---
In reply to: [191318870](https://github.com/Microsoft/web-build-tools/pull/666#discussion_r191318870) [](ancestors = 191318870)

Reply:
Yeah it matches that. I'm not sure that's a case we'd need to handle because that makes NPM barf too.